apis/hive/v1/clusterdeploymentcustomization_types.go

```diff
@@ -0,0 +1,102 @@
+package v1
+
+import (
+	conditionsv1 "github.com/openshift/custom-resource-status/conditions/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// LastApplyStatusType indicates the status of the customization on the last
+// applied cluster deployment. This is needed to for inventory sorting process to
+// avoid using same broken customization
+type LastApplyStatusType string
+
+const (
+	// LastApplySucceeded indicates that the customization
+	// worked properly on the last applied cluster deployment
+	LastApplySucceeded LastApplyStatusType = "Succeeded"
+	// LastApplyBrokenSyntax indicates that Hive failed to apply
+	// customization patches on install-config. More details would be found in
+	// Valid condition message.
+	LastApplyBrokenSyntax LastApplyStatusType = "BrokenBySyntax"
+	// LastApplyBrokenCloud indicates that cluster deployment provision has failed
+	// when used this customization. More details would be found in the Valid condition message.
+	LastApplyBrokenCloud LastApplyStatusType = "BrokenByCloud"
+	// LastApplyInstallationPending indicates that the customization patches have
+	// been successfully applied but provisioning is not completed yet.
+	LastApplyInstallationPending LastApplyStatusType = "InstallationPending"
+)
+
+// +genclient
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ClusterDeploymentCustomization is the Schema for clusterdeploymentcustomizations API
+// +kubebuilder:subresource:status
+// +k8s:openapi-gen=true
+// +kubebuilder:resource:scope=Namespaced
+type ClusterDeploymentCustomization struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ClusterDeploymentCustomizationSpec   `json:"spec"`
+	Status ClusterDeploymentCustomizationStatus `json:"status,omitempty"`
+}
+
+// ClusterDeploymentCustomizationSpec defines the desired state of ClusterDeploymentCustomization
+type ClusterDeploymentCustomizationSpec struct {
+	// InstallConfigPatches is a list of patches to be applied to the install-config
+	InstallConfigPatches []PatchEntity `json:"installConfigPatches,omitempty"`
+}
+
+// PatchEntity represent a json patch (RFC 6902) to be applied to the install-config
+type PatchEntity struct {
```

Comment:
refactoring to yaml-patch

Reply:
I wanted to replace `[]PatchEntity` with `yamlpatch.Patch`, but it doesn't have json tags so `make update` fails. So I kept this type and created the `yamlpatch.Patch` on the fly. Is there a better way to tackle this?

Reply:
Right, I was afraid of that. There's precedent for wrapping existing types to make them CRD-able -- e.g. [metav1.Duration](https://github.com/openshift/hive/blob/master/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/duration.go) -- but I've never gotten it to work for myself before. It sure would be useful in this case though...

Reply:
[Asked in #forum-api-review](https://coreos.slack.com/archives/CE4L0F143/p1651163084204559): the answer is "don't do that". So yeah, please duplicate the struct and write a simple converter we can use after we unmarshal.

Reply:
> Should we include From so that [move](https://datatracker.ietf.org/doc/html/rfc6902#section-4.4) and (probably more importantly) [copy](https://datatracker.ietf.org/doc/html/rfc6902#section-4.5) are supported?

I don't see value in `From` but I won't restrict it from users, I will add it in the next commit.

Reply:
It's required to do `move` and `copy`. I don't see a lot of utility for `move`, but I can see `copy` e.g. if I needed the generated infraID in another field somewhere. Anyway, it's part of the RFC we're claiming to support, so...